presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNIntoTableScan.java

```diff
@@ -0,0 +1,101 @@
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.TableHandle;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.connector.SortItem;
+import io.prestosql.spi.connector.SortOrder;
+import io.prestosql.spi.connector.TopNFunction;
+import io.prestosql.sql.planner.OrderingScheme;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.planner.plan.TopNNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.tableScan;
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+
+public class PushTopNIntoTableScan
+        implements Rule<TopNNode>
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+
+    private static final Pattern<TopNNode> PATTERN = topN().with(source().matching(
+            tableScan().capturedAs(TABLE_SCAN)));
+
+    private final Metadata metadata;
+
+    public PushTopNIntoTableScan(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode node, Captures captures, Context context)
+    {
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        TopNFunction topNFunction = toTopNFunction(node);
+        Map<String, ColumnHandle> assignments = tableScan.getAssignments()
+                .entrySet().stream()
+                .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+
+        Optional<TableHandle> result = metadata.applyTopN(
+                context.getSession(),
+                tableScan.getTable(),
+                topNFunction,
+                assignments);
+
+        if (result.isEmpty()) {
+            return Result.empty();
+        }
+
+        return Result.ofPlanNode(
+                        TableScanNode.newInstance(
+                                context.getIdAllocator().getNextId(),
+                                result.get(),
+                                tableScan.getOutputSymbols(),
+                                tableScan.getAssignments()));
+    }
+
+    private TopNFunction toTopNFunction(TopNNode topNNode)
+    {
+        long count = topNNode.getCount();
+        OrderingScheme orderingScheme = topNNode.getOrderingScheme();
+        List<SortItem> sortBy = orderingScheme.getOrderBy().stream()
```

Comment:
please extract `private static List<SortItem> toSortItems(OrderingScheme orderingScheme)` here

Reply:
I actually moved the method to `OrderingScheme` class as we already reuse this for both TopN and Agg Pushdown.